crates/ty_python_semantic/src/site_packages.rs

```diff
@@ -19,14 +19,55 @@ use ruff_python_ast::PythonVersion;
 
 type SitePackagesDiscoveryResult<T> = Result<T, SitePackagesDiscoveryError>;
 
+#[derive(Debug)]
+pub(crate) enum PythonEnvironment {
+    Virtual(VirtualEnvironment),
+    System(SystemEnvironment),
+}
+
+impl PythonEnvironment {
+    pub(crate) fn new(
+        path: impl AsRef<SystemPath>,
+        origin: SysPrefixPathOrigin,
+        system: &dyn System,
+    ) -> SitePackagesDiscoveryResult<Self> {
+        let path = SysPrefixPath::new(path, origin, system)?;
+
+        // Attempt to inspect as a virtual environment first
+        // TODO(zanieb): Consider avoiding the clone here by checking for `pyvenv.cfg` ahead-of-time
+        match VirtualEnvironment::new(path.clone(), origin, system) {
+            Ok(venv) => Ok(Self::Virtual(venv)),
+            // If there's not a `pyvenv.cfg` marker, attempt to inspect as a system environment
+            //
+            Err(SitePackagesDiscoveryError::NoPyvenvCfgFile(_, _)) => {
+                Ok(Self::System(SystemEnvironment::new(path)))
+            }
```

Comment:
We should probably not use this fallback if the `origin` is specific to virtual environments, I can do that in a follow-up since it's not critical.

Reply:
See https://github.com/astral-sh/ruff/pull/18003